### PR TITLE
Crackable

### PR DESCRIPTION
**Pull request by jpc20**
Added the KeyFinder class. It has the responsibility of finding the key for the crack method. I chose to make this a new class because finding the key requires a number of steps and I wanted to encapsulate it in a class of its own.
Added the module Crackable to store common behaviors for the Decryptor class and KeyFinder class.
Also added test to account for more edge cases.

### CODE REVIEW

File: lib/crackable.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  def shifts_for_less_than_4_chars(split_message, end_array, last_chars_length)
+    shifts = split_message.last.zip(end_array.last(last_chars_length)).map do |encrypted_char, end_char|
+      shift_amount(encrypted_char, end_char)
+    end
+    shifts << split_message[-2].last(4 - last_chars_length).each_with_index.map do |encrypted_char, index|
+      shift_amount(encrypted_char, end_array[index])
+    end
+    shifts.flatten
+  end
+
```
Comment (jpc20) [inline lib/crackable.rb:29]: These methods are shared by both Decryptor and KeyFinder

File: lib/decryptor.rb [hunk 2]
```diff
@@ -1,6 +1,14 @@
 require_relative "machine"
+require_relative "key_finder"
+require_relative "crackable"
 
 class Decryptor < Machine
+  include Crackable
+
+  def initialize
+    @key_finder = Keyfinder.new
+    super
```
Comment (jpc20) [inline lib/decryptor.rb:10]: Initialize an instance of KeyFinder while still inheriting from the parent.

File: lib/key_finder.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
+require_relative "machine"
+require_relative "crackable"
+
+class Keyfinder < Machine
+  include Crackable
```
Comment (jpc20) [inline lib/key_finder.rb:5]: This class is responsible for finding the key when given an encrypted message and date.

File: lib/machine.rb [hunk 4]
```diff
@@ -25,4 +25,8 @@ def split_message(message)
     end
   end
 
+  def shift_amount(encrypted_char, end_char)
+    (@alphabet.find_index(encrypted_char) - @alphabet.find_index(end_char)) % 27
+  end
```
Comment (jpc20) [inline lib/machine.rb:30]: I moved this method to the parent class because it is shared by KeyFinder and Decryptor and it uses the instance variable alphabet, so it wouldn't work in a module.

File: lib/decryptor.rb [hunk 5]
```diff
@@ -46,95 +54,8 @@ def decrypt_character(char, shift)
     @alphabet[new_index]
   end
 
-  def find_shifts(encrypted_message, date)
```
Comment (jpc20) [inline lib/decryptor.rb:49]: Moved all of these tests to the KeyFinder tests.

File: lib/decryptor.rb [hunk 6]
```diff
@@ -46,95 +54,8 @@ def decrypt_character(char, shift)
...
-  end
-
-  def shifts_minus_offsets(encrypted_message, date)
-    offsets = date_to_offsets(date)
-    find_shifts(encrypted_message, date).each_with_index.map do |shift, index|
-      shifted = shift - offsets[index]
-      shifted %= 27 if shifted < 0
-      shifted
-    end
+    @key_finder.find_key(encrypted_message, date)
```
Comment (jpc20) [inline lib/decryptor.rb:58]: All the logic behind finding the key is in the KeyFinder class.
